core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityCacheComponent.scala

```diff
@@ -0,0 +1,73 @@
+package org.broadinstitute.dsde.rawls.dataaccess.slick
+
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.rawls.monitor.EntityStatisticsCacheMonitor
+import slick.jdbc.JdbcProfile
+
+import java.sql.Timestamp
+import java.util.UUID
+
+case class EntityCacheRecord(workspaceId: UUID, entityCacheLastUpdated: Timestamp, errorMessage: Option[String])
+
+trait EntityCacheComponent {
+  this: DriverComponent =>
+
+  import driver.api._
+
+  class EntityCacheTable(tag: Tag) extends Table[EntityCacheRecord](tag, "WORKSPACE_ENTITY_CACHE") {
+    def workspaceId = column[UUID]("workspace_id", O.PrimaryKey)
+    def entityCacheLastUpdated = column[Timestamp]("entity_cache_last_updated",
+      O.SqlType("TIMESTAMP(6)"), O.Default(EntityStatisticsCacheMonitor.MIN_CACHE_TIME))
+    def errorMessage = column[Option[String]]("error_message")
+
+    def * = (workspaceId, entityCacheLastUpdated, errorMessage) <> (EntityCacheRecord.tupled, EntityCacheRecord.unapply)
+  }
+
+  object entityCacheQuery extends TableQuery(new EntityCacheTable(_)) with RawSqlQuery with LazyLogging {
+
+    val driver: JdbcProfile = EntityCacheComponent.this.driver
+
+    def findMostOutdatedEntityCacheAfter(minCacheTime: Timestamp, maxModifiedTime: Timestamp): ReadAction[Option[(UUID, Timestamp, Option[Timestamp])]] = {
+      // Find the workspace that has the entity cache that is the most out of date:
+      // A. Workspace has a cacheLastUpdated date that is not current ("current" means equal to lastModified)
+      // B. cacheLastUpdated is after @param timestamp
+      // C. lastModified is before @param cooldownBound, meaning the workspace isn't likely actively being updated
+      // D. Ordered by lastModified from oldest to newest. Meaning, return the workspace that was modified the longest ago
+
+      val baseQuery = sql"""SELECT * FROM (
+              |  SELECT w.id, w.last_modified, c.entity_cache_last_updated
+              |    FROM WORKSPACE w LEFT OUTER JOIN WORKSPACE_ENTITY_CACHE c
+              |    on w.id = c.workspace_id
+              |    where w.last_modified < $maxModifiedTime) workspacesAndCacheTimes
+              |  WHERE entity_cache_last_updated IS NULL
+              |    or ($minCacheTime < entity_cache_last_updated AND entity_cache_last_updated < last_modified)
+              |  ORDER BY last_modified asc
+              |  LIMIT 1;
+              |""".stripMargin.as[(UUID, Timestamp, Option[Timestamp])]
```

Comment:
Suggestions on how to optimize this SQL query?

Reply:
I don't believe this is necessarily any more performant, but for me it's easier to follow 
```
SELECT w.id, w.last_modified, c.entity_cache_last_updated
FROM WORKSPACE w LEFT OUTER JOIN WORKSPACE_ENTITY_CACHE c
    on w.id = c.workspace_id
where (c.entity_cache_last_updated > $minCacheTime or c.entity_cache_last_updated is null)
and w.last_modified < $maxModifiedTime and  c.entity_cache_last_updated < w.last_modified
order by w.last_modified asc limit 1
```
Probably not something to include in this PR, but the query is already doing the work of sorting all the caches that need refreshing and I think that sort is stable so you might as well just return all of them as a list and process them in scala at whatever speed you'd like and only run this query when the sweeper is out of caches to refresh.  That way we're asking less from the db.